providers/adfs_test.go

```diff
@@ -0,0 +1,147 @@
+package providers
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+
+	"github.com/coreos/go-oidc"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type fakeAdfsJwks struct{}
+
+func (fakeAdfsJwks) VerifySignature(_ context.Context, jwt string) (payload []byte, err error) {
+	decodeString, err := base64.RawURLEncoding.DecodeString(strings.Split(jwt, ".")[1])
+	if err != nil {
+		return nil, err
+	}
+	return decodeString, nil
+}
+
+func testAdfsProvider(hostname string) *ADFSProvider {
+
+	o := oidc.NewVerifier(
+		"https://issuer.example.com",
+		fakeAdfsJwks{},
+		&oidc.Config{ClientID: "https://test.myapp.com"},
+	)
+
+	p := NewADFSProvider(NewOIDCProvider(
+		&ProviderData{
+			ProviderName: "",
+			LoginURL:     &url.URL{},
+			RedeemURL:    &url.URL{},
+			ProfileURL:   &url.URL{},
+			ValidateURL:  &url.URL{},
+			Scope:        "",
+			Verifier:     o,
+		}))
+
+	if hostname != "" {
+		updateURL(p.Data().LoginURL, hostname)
+		updateURL(p.Data().RedeemURL, hostname)
+		updateURL(p.Data().ProfileURL, hostname)
+		updateURL(p.Data().ValidateURL, hostname)
+	}
+
+	return p
+}
+
+func testAdfsBackend() *httptest.Server {
+
+	authResponse := `
```

Comment:
I think this also needs a `expires_in` field to align with the OAuth spec for this response.

Both this & the `refreshResponse` need a `"token_type": "bearer"` as well assuming it follows traditional OAuth responses.